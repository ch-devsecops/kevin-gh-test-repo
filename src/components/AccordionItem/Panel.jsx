import React, { useEffect, useLayoutEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { Box, Copy } from '@honda-canada/design-system-react';
import themeStyles from './Panel.styles';
import loadLazyLoadedImages from '../../utils/loadLazyLoadedImages';
import { outerHeight } from '../../utils/DOM';

const Container = themeStyles.apply(Box, 'Container');
const Item = themeStyles.apply(Box, 'Item');
const Content = themeStyles.apply(Box, 'Content');
const Title = themeStyles.apply(Copy, 'Title');

const Panel = ({ title, id, activeTabs, setActiveTabs, margins, pageEditing, children, ...rest }) => {
  const [height, setHeight] = useState('0');
  const headerRef = useRef();
  const isActive = activeTabs?.includes(id) || pageEditing;
  const activateTab = () =>
    setActiveTabs(prevIds => (!prevIds?.includes(id) ? [...prevIds, id] : prevIds?.filter(index => index !== id)));
  const accordionPanelClassName = 'accordion-item';
  useEffect(() => {
    loadLazyLoadedImages(`.${accordionPanelClassName}`);
  }, []);

  useLayoutEffect(() => {
    if (headerRef?.current) {
      if (isActive) {
        setHeight(outerHeight(headerRef?.current));
        const timer = setTimeout(() => setHeight('none'), 200);
        return () => clearTimeout(timer);
      }
      setHeight(outerHeight(headerRef?.current));
      const timer = setTimeout(() => setHeight(0), 200);
      return () => clearTimeout(timer);
    }
  }, [isActive]);

  return (
    <Container
      margins={margins}
      className={rest?.className ? `${rest.className} ${accordionPanelClassName}` : `${accordionPanelClassName}`}
      {...rest}
    >
      <Item
        data-testid="accordion"
        onClick={activateTab}
        role="button"
        aria-expanded={isActive}
        isEditorActive={pageEditing}
      >
        <Title>{title}</Title>
      </Item>
      <Content
        data-testid="accordion-content"
        ref={headerRef}
        height={height}
        aria-hidden={!isActive}
        isActive={isActive}
      >
        {children}
      </Content>
    </Container>
  );
};

Panel.propTypes = {
  title: PropTypes.string,
  margins: PropTypes.shape({
    autoId: PropTypes.string,
    innerBottomMargin: PropTypes.string,
    innerBottomMarginMob: PropTypes.string,
    innerHorzMargin: PropTypes.string,
    innerHorzMarginMob: PropTypes.string,
    innerTopMargin: PropTypes.string,
    innerTopMarginMob: PropTypes.string,
  }),
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
  setActiveTabs: PropTypes.func.isRequired,
  activeTabs: PropTypes.arrayOf(PropTypes.string),
  id: PropTypes.string.isRequired,
};

export default Panel;
