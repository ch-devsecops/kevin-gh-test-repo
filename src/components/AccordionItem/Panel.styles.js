import css from '@styled-system/css';
import { ThemeStyles } from '@honda-canada/design-system-react';

const themeStyles = new ThemeStyles();

themeStyles.addBaseStyles('Container')(({ theme, margins = {} }) =>
  css({
    borderBottom: theme.borders[2],
    borderColor: 'grey.3',
    '[aria-expanded="true"]:before': {
      transform: 'rotate(0deg)',
    },
    '[aria-hidden="true"]': {
      mt: 0,
      mb: 0,
      mx: [margins?.innerHorzMarginMob, margins?.innerHorzMarginMob, margins?.innerHorzMargin],
    },
    '[aria-hidden="false"]': {
      mt: [margins?.innerTopMarginMob, margins?.innerTopMarginMob, margins?.innerTopMargin],
      mb: [margins?.innerBottomMarginMob, margins?.innerBottomMarginMob, margins?.innerBottomMargin],
      mx: [margins?.innerHorzMarginMob, margins?.innerHorzMarginMob, margins?.innerHorzMargin],
    },
  }),
);

themeStyles.addBaseStyles('Item')(({ theme, isEditorActive }) =>
  css({
    position: 'relative',
    width: '100%',
    pl: ['15px', 'default', 'l'],
    pr: 'xl',
    py: 'zero',
    transition: 'color 0.2s linear',
    cursor: 'pointer',

    '&::after, &::before': {
      width: theme.space.m,
      backgroundColor: 'black',
      content: '""',
      position: 'absolute',
      right: ['11px', 'm'],
      top: '50%',
      height: '2px',
      marginTop: '-2px',
    },

    '&::before': {
      transform: 'rotate(-90deg)',
      transition: 'transform 200ms ease-in-out 0s',
    },
    minHeight: isEditorActive ? theme.space[7] : [theme.space[6], theme.space[8]],
    pt: isEditorActive ? 'l' : undefined,
  }),
);

themeStyles.addBaseStyles('Title')(
  css({
    fontFamily: 'bold',
    pointerEvents: 'auto',
    pb: ['xs', 'default', 'l'],
    pt: ['s', 'default', 'l'],
    lineHeight: ['1.3', 'sm'],
    fontSize: ['m', '20px'],
  }),
);

themeStyles.addBaseStyles('Content')(({ isActive }) =>
  css({
    overflow: 'hidden',
    willChange: 'height',
    transition: 'margin ease 300ms, height ease 300ms',
    '& > *': {
      transition: 'visibility 400ms',
      visibility: !isActive && 'hidden',
    },
  }),
);

export default themeStyles;
