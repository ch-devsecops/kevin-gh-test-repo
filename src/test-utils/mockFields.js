/* eslint-disable max-len */
/* eslint-disable max-classes-per-file */

// FIXME: jsdoc popup not working on test files
/** Class representing a mock data. */
class MockData {
  /**
   * Add a field.
   * @param {string} key - key of the field obj: i.e. 'title'
   * @param {Object} value - value of the field obj: i.e. { value: 'some text' }.
   */
  add(key, value) {
    this[key] = value;
  }
}

/**
 * Class representing mock fields object.
 * @extends MockData
 *
 * available method: add()
 *
 * default fields: [anchorId, title, bodyText, image, ctaLink, ctaType, ctaIcon, gtmCategory, gtmTitle, contentTitle, subtitle]
 *
 * note: we can add more commonly used fields in the class declaration if needed
 */
class MockFields extends MockData {
  constructor() {
    super();

    this.title = { value: 'Title' };
    this.bodyText = { value: 'BodyText' };
    this.image = {
      value: { src: 'mock.png', alt: 'mock alt' },
    };
    this.ctaLink = {
      value: {
        href: 'www.source.ca',
        title: 'cta aria label',
        text: 'CTA Text',
      },
    };
    this.ctaType = {
      value: 'Tertiary',
    };
    this.ctaIcon = {
      fields: {
        value: {
          value: 'arrowRight',
        },
      },
    };
    this.gtmCategory = {
      fields: {
        value: {
          value: 'Others',
        },
      },
    };
    this.gtmTitle = {
      value: 'gtm title',
    };
    this.contentTitle = {
      value: 'Content Title',
    };
    this.subtitle = {
      value: 'Content Subtitle',
    };
    this.seoH1 = {
      value: '',
    };
    this.anchorId = {
      value: 'anchor-id',
    };
    this.ctaLabel = {
      value: 'CTA Label',
    };
  }
}

/**
 * Class representing a mock rendering object
 * @extends MockData
 *
 * available method: add()
 */
class MockRendering extends MockData {
  /**
   * Create a mock rendering object
   * @param {string} componentName - adds componentName
   */
  constructor(componentName) {
    super();

    this.componentName = componentName;
  }
}

export { MockFields, MockRendering };
