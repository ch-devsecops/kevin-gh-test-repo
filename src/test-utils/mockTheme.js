/**
 * Mock styled-components theme, if you want to test how your component
 * implements theme styles. Import as-is, or spread in any
 * additional values you need.
 *
 * Example usage:

  it('should apply the bgColor param', () => {
    const { colors } = MockTheme;
    const { container } = render(
      <MyComponent
        fields={mockFields}
        params={{ bgColor: 'Gray' }}
      />,
    );
    const wrapper = container.querySelector('section');

    expect(wrapper).toHaveStyleRule('background-color', colors.grey[5]);
  });
 */

import { HONDA_THEME_NAME } from '../utils/constants';

const mockTheme = {
  letterSpacings: {
    sm: '0.5px',
    md: '1px',
  },
  lineHeights: {
    sm: '22px',
    md: '26px',
  },
  colors: {
    white: '#fff',
    transparent: 'transparent',
    grey: ['#707070', '#C7C7C7', '#D5D5D5', '#EAEAEA', '#F8F8F8', '#F5F5F5'],
    black: '#2C2C2C',
    red: '#DA2C2E',
    redShades: ['rgba(200, 16, 46, 0.1)'],
    aodaRed: '#D92629',
    lightRed: '#EC9597',
    darkRed: '#A21C1D',
    blue: '#334C96',
    darkBlue: '#002D62',
    lightBlue: '#00447826',
    yellow: ['#F9CA52', '#EDAB00', '#8D763F'],
    fadedBlue: '#ADB7D5',
    primary: '#DA2C2E',
    primaryAoda: '#D92629',
    primaryDark: '#A21C1D',
    primaryLight: '#EC9597',
    secondary: '#2C2C2C',
    secondaryLight: '#C7C7C7',
    tertiary: '#002D62',
    tertiaryLight: '#334C96',
    tertiaryDisabled: '#ADB7D5',
    danger: '#DA2C2E',
    warning: '#2C2C2C',
    warningLight: '#EAEAEA',
    success: '#334C96',
    successLight: '#d9e0e7',
    disabled: '#C7C7C7',
    typographyDefault: '#2C2C2C',
  },
  space: [0, 4, 8, 12, 16, 24, 32, 48, 64, 96, 128, 144, 192, 256],
  borders: [0, '1px solid', '2px solid', '3px solid'],
  breakpoints: ['768px', '992px', '1200px'],
  fontWeights: {
    regular: 'normal',
    bold: 700,
  },
  fontSizes: {
    sm: '12px',
  },
  shadows: {
    smooth: {
      md: '0.1px 0.3px 0.3px rgba(0, 0, 0, 0.02),0.3px 0.7px 0.9px rgba(0, 0, 0, 0.028),0.7px 1.4px 1.8px rgba(0, 0, 0, 0.035),1.5px 2.9px 3.7px rgba(0, 0, 0, 0.047),4px 8px 10px rgba(0, 0, 0, 0.07)',
    },
    box: {
      subNav: '0px 2px 4px rgba(0, 0, 0, 0.1)',
    },
  },
  zIndices: {
    accordionItemTitle: 200,
    modal: 13001,
    sectionNav: 13000,
  },
  transitionDuration: {
    t1: '1s',
  },
  transitionTimingFunction: {
    default: 'ease-in-out',
    out: 'ease-out',
    in: 'ease-in',
  },
  name: HONDA_THEME_NAME,
  fonts: {
    default: 'Helvetica',
    bold: 'Cooper Black',
    heading: 'Roboto',
  },
  header: {
    desktop: {},
    mobile: {},
  },
  icons: {},
  logIconWarnings: false,
};

export default mockTheme;
