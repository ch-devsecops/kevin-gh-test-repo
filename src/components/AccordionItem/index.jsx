import React from 'react';
import { Placeholder } from '@sitecore-jss/sitecore-jss-react';
import PropTypes from 'prop-types';
import Panel from './Panel';
import getContentMargins from '../../utils/getContentMargins';

const AccordionItem = ({ rendering, gtmTags, activeTabs, setActiveTabs, gtmEventPayload, pageEditing, appName }) => {
  const accordionTitle = rendering?.fields?.title?.value;
  const margins = getContentMargins(rendering?.params);

  return (
    <Panel
      id={rendering?.uid}
      activeTabs={activeTabs}
      setActiveTabs={setActiveTabs}
      title={accordionTitle}
      margins={margins}
      pageEditing={pageEditing}
      appName={appName}
      gtmEventPayload={gtmEventPayload}
      {...gtmTags}
    >
      <Placeholder name="accordion-item-content" rendering={rendering} />
    </Panel>
  );
};

AccordionItem.propTypes = {
  rendering: PropTypes.shape({
    params: PropTypes.shape({
      innerTopMargin: PropTypes.string,
      innerBottomMargin: PropTypes.string,
      innerHorzMargin: PropTypes.string,
      innerTopMarginMob: PropTypes.string,
      innerBottomMarginMob: PropTypes.string,
      innerHorzMarginMob: PropTypes.string,
    }),
    placeholders: PropTypes.shape({ 'accordion-item-content': PropTypes.arrayOf(PropTypes.shape({})) }),
    fields: PropTypes.shape({ title: PropTypes.shape({ value: PropTypes.string }) }),
    uid: PropTypes.string,
  }),
  gtmTags: PropTypes.shape({
    gtmTitle: PropTypes.string,
    trimName: PropTypes.string,
    bodyStyle: PropTypes.string,
    modelName: PropTypes.string,
    componentName: PropTypes.string,
    interactionType: PropTypes.string,
  }),
  activeTabs: PropTypes.arrayOf(PropTypes.string),
  setActiveTabs: PropTypes.func,
};

export default AccordionItem;
