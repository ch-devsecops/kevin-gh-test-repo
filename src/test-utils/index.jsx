import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { SitecoreContext } from '@sitecore-jss/sitecore-jss-react';
import { BrowserRouter } from 'react-router-dom';
import { OptionalVideoModalContextProvider } from '@honda-canada/design-system-react';
import UserLocationProvider from '@honda-canada/user-location';
import { configureAxe } from 'jest-axe';
import ModelExplorationContextProvider from '../components/ModelExplorationContext';
import ModelFiltersContextProvider from '../components/ModelFiltersContext';
import { MockFields, MockRendering } from './mockFields';
import MockTheme from './mockTheme';
import 'intersection-observer';
import { ACURA_THEME_NAME } from '../utils/constants';

const generateJssWrapper = context =>
  // eslint-disable-next-line react/display-name
  function ({ children }) {
    return (
      <SitecoreContext
        context={{
          language: 'en',
          site: { name: ACURA_THEME_NAME },
          ...context?.sitecoreContext,
        }}
        componentFactory={{}}
      >
        <BrowserRouter>
          <UserLocationProvider>
            <ThemeProvider theme={context?.theme || MockTheme}>
              <OptionalVideoModalContextProvider>
                <ModelExplorationContextProvider>
                  <ModelFiltersContextProvider>{children}</ModelFiltersContextProvider>
                </ModelExplorationContextProvider>
              </OptionalVideoModalContextProvider>
            </ThemeProvider>
          </UserLocationProvider>
        </BrowserRouter>
      </SitecoreContext>
    );
  };
/**
 * A custom implementation of the testing-library `render` function that
 * wraps tests in an typical JSS app's context providers.
 * Note that you can include a `context` object to set or override values
 * for SitecoreContext and ThemeContext.
 * @param {ReactElement} component - the component you want testing-library to render
 * @param {object} [options] - see https://testing-library.com/docs/react-testing-library/api/#render-options
 * @param {string} [options.context] - default values for SitecoreContext, ThemeProvider
 * @param {string} [options.context.sitecoreContext] - value for SitecoreContext
 * @param {string} [options.context.theme] - value for ThemeProvider
 * @example
 * const { getByText } = jssRender(
 *  <MyComponent />, { context: { sitecoreContext: { language: 'fr' } } }
 * );
 * @returns {function(): void} custom testing-library render function
 */
const jssRender = (component, options = {}) => {
  const { context } = options;
  // eslint-disable-next-line no-param-reassign
  delete options.context;

  return render(component, {
    wrapper: generateJssWrapper(context),
    ...options,
  });
};

const generateWrapper = context =>
  // eslint-disable-next-line react/display-name
  function ({ children }) {
    return (
      <BrowserRouter>
        <ThemeProvider theme={context?.theme || MockTheme}>
          <OptionalVideoModalContextProvider>{children}</OptionalVideoModalContextProvider>
        </ThemeProvider>
      </BrowserRouter>
    );
  };

/**
 * A custom implementation of the testing-library `render` function that
 * wraps tests in typical design-system context providers.
 * Note that you can include a `theme` object to set or override values
 * for ThemeContext.
 * @param {ReactElement} component - the component you want testing-library to render
 * @param {object} [options] - see https://testing-library.com/docs/react-testing-library/api/#render-options
 * @param {string} [options.theme] - default value for ThemeProvider
 * @example
 * const { getByText } = render(
 *  <MyComponent />, { theme: { space: { wrapperGap: '15px' } } }
 * );
 * @returns {function(): void} custom testing-library render function
 */
const customRender = (component, options = {}) => {
  const { theme } = options;
  // eslint-disable-next-line no-param-reassign
  delete options.theme;

  return render(component, {
    wrapper: generateWrapper(theme),
    ...options,
  });
};

// disable accessibility checks for errors that are usually trigged by content mistakes
const axe = configureAxe({
  rules: {
    'duplicate-id': { enabled: false },
    'heading-order': { enabled: false },
  },
});

export * from 'jest-axe';
export { axe };
export * from '@testing-library/react';
export { customRender as render, jssRender };
export { MockFields, MockRendering, MockTheme };
