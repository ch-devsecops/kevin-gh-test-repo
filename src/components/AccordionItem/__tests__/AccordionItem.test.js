import React from 'react';
import { axe, cleanup, jssRender } from 'test-utils';
import AccordionItem from '..';

afterEach(cleanup);

const mockData = require('../../../../sample-mock-data/routes/accordion-item/en.json');
const emptyMockData = require('../../../../sample-mock-data/routes/accordion-item/empty_en.json');

describe('<AccordionItem>', () => {
  test('should render correctly', () => {
    const { container } = jssRender(<AccordionItem {...mockData} />);

    expect(container).toMatchSnapshot();
    expect(container.children[0]).toBeTruthy();
  });

  test('should return null if there is no fields', () => {
    const { container } = jssRender(<AccordionItem fields={null} />);

    expect(container).toMatchSnapshot();
  });

  test('should return null if the data is empty', () => {
    const { container } = jssRender(<AccordionItem {...emptyMockData} />);
    expect(container).toMatchSnapshot();
  });

  test('should have no accessibility violations', async () => {
    const renderingMock = {
      fields: {
        title: {
          value: 'Western and Pacific Canada',
        },
      },
    };
    const { container } = jssRender(<AccordionItem {...mockData} rendering={renderingMock} />);

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  });
});
