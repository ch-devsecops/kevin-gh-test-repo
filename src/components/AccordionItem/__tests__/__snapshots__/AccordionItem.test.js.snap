// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccordionItem> should have no accessibility violations 1`] = `
<div>
  <div
    class="Box-sc-1bsf784-0 sc-1k44dfv-0 fYkisg accordion-item"
  >
    <div
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 hfklgf"
      data-testid="accordion"
      role="button"
    >
      <p
        class="sc-1k44dfv-0 sc-1k44dfv-0 dfqSgi"
        color="typographyDefault"
        font-family="default"
        font-weight="regular"
      >
        Western and Pacific Canada
      </p>
    </div>
    <div
      aria-hidden="true"
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 kTXKFc"
      data-testid="accordion-content"
      height="0"
    />
  </div>
</div>
`;

exports[`<AccordionItem> should render correctly 1`] = `
<div>
  <div
    class="Box-sc-1bsf784-0 sc-1k44dfv-0 fYkisg accordion-item"
  >
    <div
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 hfklgf"
      data-testid="accordion"
      role="button"
    >
      <p
        class="sc-1k44dfv-0 sc-1k44dfv-0 dfqSgi"
        color="typographyDefault"
        font-family="default"
        font-weight="regular"
      />
    </div>
    <div
      aria-hidden="true"
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 kTXKFc"
      data-testid="accordion-content"
      height="0"
    />
  </div>
</div>
`;

exports[`<AccordionItem> should return null if the data is empty 1`] = `
<div>
  <div
    class="Box-sc-1bsf784-0 sc-1k44dfv-0 fYkisg accordion-item"
  >
    <div
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 hfklgf"
      data-testid="accordion"
      role="button"
    >
      <p
        class="sc-1k44dfv-0 sc-1k44dfv-0 dfqSgi"
        color="typographyDefault"
        font-family="default"
        font-weight="regular"
      />
    </div>
    <div
      aria-hidden="true"
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 kTXKFc"
      data-testid="accordion-content"
      height="0"
    />
  </div>
</div>
`;

exports[`<AccordionItem> should return null if there is no fields 1`] = `
<div>
  <div
    class="Box-sc-1bsf784-0 sc-1k44dfv-0 fYkisg accordion-item"
  >
    <div
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 hfklgf"
      data-testid="accordion"
      role="button"
    >
      <p
        class="sc-1k44dfv-0 sc-1k44dfv-0 dfqSgi"
        color="typographyDefault"
        font-family="default"
        font-weight="regular"
      />
    </div>
    <div
      aria-hidden="true"
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 kTXKFc"
      data-testid="accordion-content"
      height="0"
    />
  </div>
</div>
`;
